design/design-doc-composition.md

```diff
@@ -0,0 +1,1079 @@
+  # resources. This allows the use of RBAC to restrict which resources a
+  # MySQLInstance may be composed of. The specified service account must have
+  # full access to MySQLInstance resources, and 'get' access to Component
+  # resources.
+  #
+  # If the service account is omitted Crossplane will use its pod service
+  # account to manage MySQLInstance resources. This implies that anyone with
+  # sufficient RBAC permissions to create a Composition and to create a
+  # MySQLInstance will be able to compose their MySQLInstance of any
+  # infrastructure resource that Crossplane is able to create.
+  serviceAccountRef:
+    namespace: crossplane-system
+    name: wordpresses.apps.example.org
+```
+
+When an application developer authors the above `kind: InfrastructureDefinition`
+Crossplane will automatically create a `kind: CustomResourceDefinition`, that
+allows application operators to author the below custom resource:
+
+```yaml
+apiVersion: database.example.org/v1alpha1
+kind: MySQLInstance
+metadata:
+  name: sql
+spec:
+  # The schema for the following three fields is defined by the above
+  # InfrastructureDefinition.
+  engineVersion: "5.7"
+  storageGB: 10
+  region: us-west
+  # The infrastructure object schema is automatically injected into the
+  # CustomResourceDefinition that is created by the InfrastructureDefinition
+  # that defines the MySQLInstance resource.
+  infrastructure:
+    # Multiple compositions may potentially satisfy a particular kind of
+    # infrastructure. Exactly one composition may be annotated as the default
+    # for a particular kind of infrastructure. Each infrastructure instance may
+    # influence which composition is used via label selectors. This could be
+    # used, for example, to determine whether a GCP CloudSQLInstance or an Azure
+    # SQLServer based composition satisfied this MySQLInstance.
+    compositionSelector:
+     matchLabels:
+       connectivity: private
+    # The MySQLInstance author may explicitly select which composition should be
+    # used by setting the compositionRef. In the majority of cases the author
+    # will ignore this field and it will be set by a controller, similar to the
+    # contemporary classRef field.
+    compositionRef:
+    - name: private-mysql-server
+    # Each infrastructure resource maintains an array of the resources it
+    # composes. Composed resources are always cluster scoped, and always either
+    # primitive or composite infrastructure resources. Composed resources model
+    # their relationship with the infrastructure resource via their controller
+    # reference. The infrastructure resource must maintain this array because
+    # there is currently no user friendly, performant way to discover which
+    # resources (of arbitrary kinds) are controlled by a particular resource per
+    # https://github.com/kubernetes/kubernetes/issues/54498
+    composedRefs:
+    - apiVersion: azure.crossplane.io/v1alpha3
```

Comment:
I think it's worth to talk about how this relation is represented on the composed resources, too.

The requirements of the relation are:
* The managed resource/composition instance should not get bound to another composition instance and/or claim.
  * This has to consider old claims, too, since we're keeping them.
* After restoration of a backup, owner references are stripped out.
  * Composition instance has a list of resources, so, it knows which resources it should managed.
  * The primitive resources should not allow any other instance/old claim to bind or manage it. If the whole relation was built on owner ref, then that's not possible since during restoration there might be claims/instances newly created without any resource ref, so, they would bind to those resources.
* Relation might use uid but it should be flexible enough to work without uid, like secret propagation that was revised for backup/restore.
* It should consider the nested cases where a composed cluster-scoped instance has another composed cluster-scoped instance as composed resource. So, it should not make any difference whether the user used managed resource or a cluster-scoped instance.

Binding phase doesn't work for this since it's in status. Only owner reference won't work since owner references are stripped after backup/restore process but the reason owner reference is not enough is that there isn't any way to prevent other parent-like resource to add their ownerRefs to those resources after restoration until actual parent does. But if that's the only thing, we can actually solve this via a field on spec like `bound: true/false`. So, here is what I have in mind for cluser-scoped instance and managed resources:

* The parent to child relation will be represented as `spec.resourceRefs[]` or `spec.composedRefs[]`.

* The child to parent relation will be represented as `metadata.ownerRef` + `spec.bound: true`

This is true for both cluster-scoped<>primitive managed and cluster-scoped<>cluster-scoped instance.

For namespaced, we can't use owner ref since namespaced cannot own cluster-scoped and the relation is 1-to-1:

* Namespaced has `spec.resourceRef`
* Cluster-scoped has `spec.bound: true`, no owner reference.


The only addition will be that we will have a `spec.bound: true/false` on all composable resources (managed resources and cluster-scoped instances) as addition to owner reference and this will allow us to remove the `spec.claimRef`-like field to refer to the parent resource since owner ref will do that already.

Reply:
> This has to consider old claims, too, since we're keeping them.

True, but only temporarily. I imagine we'll deprecate them.

> The primitive resources should not allow any other instance/old claim to bind or manage it.

> there isn't any way to prevent other parent-like resource to add their ownerRefs to those resources after restoration until actual parent does.

Is this really a problem? I believe for one parent to steal another's child we'd need two parent resources to already be racing to own one child resource. Presuming potential parent A won originally, is it so bad that we let potential parent B take ownership after a backup and restore? We end up in effectively the same position after the restore; with two parents fighting to own one child, but only one owning it. This aligns with the [controller ref design doc](https://github.com/kubernetes/community/blob/ead3b8e5ec554217e8a4935112462b98ff28fecb/contributors/design-proposals/api-machinery/controller-ref.md), and seems to simplify our design.

Reply:
Alright, I think I confused somethings here when I had the implementation with binding phase. Right now, I removed the binding phase and composite resource won't take over any existing composed resource unless it has the composed resource's reference in `spec.resourceRefs[]`, which eliminates the cases where multiple composites trying to take over a composed resource. That's the intent, right? Static provisioning will happen only in namespaced<>composite level.

I was thinking that since the composed resource won't have any lock info after restoration, composites w/o `resourceRefs[]` will race with all the ones that have but since there is no take-over, there is no race.